### ISSUE TITLE
New lint: duplicate method

### ISSUE THREAD

**Issue by alice-i-cecile** [rust-lang/rust-clippy#13399]
### What it does

Detects if two methods on the same struct are identical. The requirements need to be quite strict:

- trait methods don't count: defaulted impls with the same body are fine and trait objects means exposing multiple copies of the same code can be useful
- the type signature needs to be identical, otherwise mutable and immutable getters will trigger false positives

Ideally, variable renames shouldn't matter. Similarly, calling another method on the same struct (as in https://github.com/bevyengine/bevy/pull/15234) should be detected.

### Advantage

API duplication is confusing to users, annoying to maintain, and often indicative of the same bug.

### Drawbacks

This lint may be annoying during development (in the same way as dead_code), due to copy-paste-modify workflows.

Actually implementing 

### Example

```rust
struct Counter(i8);

impl Counter {
   fn get(&self) -> i8 {
       self.0
   }
   
   fn value(&self) -> i8 {
       self.0
   }
}
```

Could be written as:

```rust
struct Counter(i8);

impl Counter {
   fn get(&self) -> i8 {
       self.0
   }   
}
```

**Comment by alice-i-cecile** [comment]
You could also extend this to free-standing functions with the same body within the same namespace actually. Might be a bit more complex to check for and less common, but it fits in well.

**Comment by alice-i-cecile** [comment]
Evidently! If only we had such a lint for the `clippy` repo itself!
